### PR TITLE
Azure extension

### PR DESCRIPTION
**Pull request by codemnky**
Installer for Azure extension. This will add configuration and create several Azure specific resources for  Jazz to use in deployment: an EventHub and Resource Groups.

### CODE REVIEW

File: Installer.sh [hunk 1]
```diff
@@ -18,7 +18,7 @@ INSTALL_DIR=`pwd`
 # Log file to record the installation logs
 LOG_FILE_NAME=installer_setup.out
 LOG_FILE=`realpath $INSTALL_DIR/$LOG_FILE_NAME`
-JAZZ_BRANCH="master"
+JAZZ_BRANCH="azure-master"
```
Comment (devsatishm) [inline Installer.sh]: Revert the change.

File: installscripts/jazz-terraform-unix-noinstances/variables.tf [hunk 2]
```diff
@@ -3,7 +3,10 @@
 #
 variable "region" { type = "string" default = "us-east-1" }
 variable "github_branch" { type = "string" default = "development" }
-variable "github_repo" { type = "string" default = "https://github.com/tmobile/jazz.git" }
+variable "github_repo" {
```
Comment (devsatishm) [inline installscripts/jazz-terraform-unix-noinstances/variables.tf]: Revert the change.

File: installscripts/cookbooks/jenkins/attributes/default.rb [hunk 3]
```diff
@@ -14,7 +14,7 @@
 default['region'] = 'us-east-1'
 default['git_branch'] = 'master'
 default['git_plugin_branch'] = 'master'
-default['git_repo'] = 'https://github.com/tmobile/jazz.git'
```
Comment (devsatishm) [inline installscripts/cookbooks/jenkins/attributes/default.rb]: Revert the change.

File: feature-extensions/azure/terraform/azure.tf [hunk 4]
```diff
@@ -0,0 +1,102 @@
+# Create a role for the Lambda function to exec under
```
Comment (devsatishm) [inline feature-extensions/azure/terraform/azure.tf]: Remove/update comment.

File: feature-extensions/azure/terraform/azure.tf [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  sku = "Basic"
+  capacity = 1
+  kafka_enabled = false
+
+  tags {
+    environment = "Production"
+  }
+}
+
+resource "azurerm_eventhub" "production" {
```
Comment (devsatishm) [inline feature-extensions/azure/terraform/azure.tf]: Why do we need an event hub?

File: feature-extensions/azure/jenkins_config.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    add_jenkins_credential(jenkins_host, jenkins_port, basic_auth, 'AZ_PASSWORD', azure_client_secret)
+    add_jenkins_credential(jenkins_host, jenkins_port, basic_auth, 'AZ_SUBSCRIPTIONID', azure_subscription_id)
+    add_jenkins_credential(jenkins_host, jenkins_port, basic_auth, 'AZ_TENANTID', azure_tenant_id)
+
+
+def add_jenkins_credential(jenkins_host, jenkins_port, basic_auth, key, value):
+    url = "http://{}:{}/credentials/store/system/domain/_/createCredentials".format(jenkins_host,
+                                                                                    jenkins_port)
+    content = """{{
+  "": "0",
```
Comment (devsatishm) [inline feature-extensions/azure/jenkins_config.py]: Fix formatting.

File: feature-extensions/azure/setup.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    OKGREEN = '\033[92m'
+    WARNING = '\033[93m'
+    FAIL = '\033[91m'
+    ENDC = '\033[0m'
+    BOLD = '\033[1m'
+    UNDERLINE = '\033[4m'
+
+
+featureName = "Azure Deployment"
+configFolder = "module_config"
```
Comment (devsatishm) [inline feature-extensions/azure/setup.py]: `installer-vars.json `is usually available in `./jazz-build-module/`. What's the purpose of this location?

File: feature-extensions/azure/terraform/azure.tf (hunk 5 above)
Comment (codemnky) [reply feature-extensions/azure/terraform/azure.tf]: The event hub is used for the monitoring.

File: feature-extensions/azure/jenkins_config.py (hunk 6 above)
Comment (codemnky) [reply feature-extensions/azure/jenkins_config.py]: Replaced with non-multiline string and formatted.

File: feature-extensions/azure/setup.py (hunk 7 above)
Comment (codemnky) [reply feature-extensions/azure/setup.py]: This is just a temporary folder that holds the configuration files. It is part of jazz-build-module. Would you prefer that I name it jazz-build-module?

**Comment by rajeevr2715** [comment]
Please use code style validation using Flake8 - Python

File: feature-extensions/azure/setup.py (hunk 7 above)
Comment (devsatishm) [reply feature-extensions/azure/setup.py]: If this is a temp location, that should be fine.

File: feature-extensions/azure/setup.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+@click.option('--azure-client-secret', envvar='AZURE_CLIENT_SECRET', help='Specify the password for Service Principal',
+              prompt=True)
+@click.option('--azure-tenant-id', envvar='AZURE_TENANT_ID',
+              help='Specify the Azure AD tenant id for the Service Principal', prompt=True)
+@click.option('--azure-company-name', envvar='AZURE_COMPANY_NAME',
+              help='Specify the company name used in the Azure API Management service',
+              prompt=True)
+@click.option('--azure-company-email', envvar='AZURE_COMPANY_EMAIL',
+              help='Specify the company contact email used in the Azure API Management service', prompt=True)
+def uninstall(jazz_stackprefix, scm_repo, scm_username, scm_password, scm_pathext, azure_subscription_id,
```
Comment (bleggett) [inline feature-extensions/azure/setup.py]: Probably can't be helped, but this is a poopton of CLI args, is there nothing at all here that can be derived instead of supplied?

File: feature-extensions/azure/setup.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+    update_config(azure_subscription_id, azure_location, azure_client_id, azure_client_secret, azure_tenant_id)
+    jenkins_config.update_jenkins(jenkins_user, jenkins_host, jenkins_port, jenkins_api_token, azure_client_id,
+                                  azure_client_secret,
+                                  azure_tenant_id, azure_subscription_id)
+    commit_config("Adding Azure deployment feature")
+
+
+@main.command()
+@click.option('--jazz-stackprefix',
```
Comment (bleggett) [inline feature-extensions/azure/setup.py]: Oh shoot didn't know about the `envvar` option, stealing that for my rewrite.

File: feature-extensions/azure/setup.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+def commit_config(message):
+    git_config.commit_git_config(configFolder, jsonConfigFile, message)
+
+
+def update_config(azure_subscription_id, azure_location, azure_client_id, azure_client_secret, azure_tenant_id):
+    with open("{}/{}".format(configFolder, jsonConfigFile), 'r') as f:
+        data = json.load(f, object_pairs_hook=OrderedDict)
+
+    azureConfig = OrderedDict()
+    azureConfig['SUBSCRIPTION_ID'] = 'AZ_SUBSCRIPTIONID'
```
Comment (devsatishm) [inline feature-extensions/azure/setup.py]: We'll need `azureConfig['IS_ENABLED'] = true` so that admin can turn on/off azure features .
